File: go-client/internal/storage/filestore.go
package storage

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"golang.org/x/crypto/hkdf"

	"github.com/cisc468/p2p-project/internal/crypto"
)

const fileKeyInfo = "p2pshare-file-v1"

// DeriveFileKey derives a per-file AES-256-GCM key from the vault key and file_id.
func DeriveFileKey(vaultKey []byte, fileID string) ([]byte, error) {
	fileIDBytes, err := crypto.B64URLDecode(fileID)
	if err != nil {
		return nil, err
	}
	hkdfReader := hkdf.New(sha256.New, vaultKey, fileIDBytes, []byte(fileKeyInfo))
	key := make([]byte, 32)
	if _, err := io.ReadFull(hkdfReader, key); err != nil {
		return nil, err
	}
	return key, nil
}

// ComputeFileID computes file_id = base64url(SHA256(ownerID || utf8(filename) || timestamp_bytes)).
func ComputeFileID(ownerIDBytes []byte, filename string, uploadTimestamp int64) string {
	tsBytes := make([]byte, 8)
	binary.BigEndian.PutUint64(tsBytes, uint64(uploadTimestamp))
	h := sha256.New()
	h.Write(ownerIDBytes)
	h.Write([]byte(filename))
	h.Write(tsBytes)
	return crypto.B64URLEncode(h.Sum(nil))
}

// ComputeFileHash computes base64url(SHA256(data)).
func ComputeFileHash(data []byte) string {
	h := sha256.Sum256(data)
	return crypto.B64URLEncode(h[:])
}

// Manifest represents a signed file manifest entry.
type Manifest struct {
	FileID          string `json:"file_id"`
	Filename        string `json:"filename"`
	SizeBytes       int64  `json:"size_bytes"`
	SHA256          string `json:"sha256"`
	OwnerID         string `json:"owner_id"`
	UploadTimestamp int64  `json:"upload_timestamp"`
	Description     string `json:"description"`
	Version         int    `json:"version"`
	OwnerSignature  string `json:"owner_signature,omitempty"`
}

// ManifestWithoutSig returns the manifest fields as a map without owner_signature for signing.
func ManifestWithoutSig(m *Manifest) map[string]interface{} {
	return map[string]interface{}{
		"file_id":          m.FileID,
		"filename":         m.Filename,
		"size_bytes":       float64(m.SizeBytes),
		"sha256":           m.SHA256,
		"owner_id":         m.OwnerID,
		"upload_timestamp": float64(m.UploadTimestamp),
		"description":      m.Description,
		"version":          float64(m.Version),
	}
}

// CreateManifest creates a signed manifest for a file.
func CreateManifest(identitySK ed25519.PrivateKey, identityPK ed25519.PublicKey,
	filename string, fileData []byte, description string) (*Manifest, error) {

	ownerIDBytes := crypto.PublicKeyToBytes(identityPK)
	uploadTS := time.Now().UnixMilli()
	fileID := ComputeFileID(ownerIDBytes, filename, uploadTS)
	fileHash := ComputeFileHash(fileData)

	m := &Manifest{
		FileID:          fileID,
		Filename:        filename,
		SizeBytes:       int64(len(fileData)),
		SHA256:          fileHash,
		OwnerID:         crypto.PeerIDFromPublicKey(identityPK),
		UploadTimestamp: uploadTS,
		Description:     description,
		Version:         1,
	}

	sigData, err := crypto.CanonicalJSON(ManifestWithoutSig(m))
	if err != nil {
		return nil, err
	}
	sig := crypto.Sign(identitySK, sigData)
	m.OwnerSignature = crypto.B64URLEncode(sig)
	return m, nil
}

// VerifyManifest verifies the owner_signature on a manifest entry.
func VerifyManifest(m *Manifest, ownerPK ed25519.PublicKey) error {
	sig, err := crypto.B64URLDecode(m.OwnerSignature)
	if err != nil {
		return fmt.Errorf("decode signature: %w", err)
	}
	sigData, err := crypto.CanonicalJSON(ManifestWithoutSig(m))
	if err != nil {
		return err
	}
	return crypto.Verify(ownerPK, sigData, sig)
}

// VerifyFileIntegrity checks that file data matches the expected hash.
func VerifyFileIntegrity(fileData []byte, expectedHashB64 string) error {
	actual := ComputeFileHash(fileData)
	if actual != expectedHashB64 {
		return errors.New("downloaded file is corrupted or was tampered with — file discarded")
	}
	return nil
}

// FileStore manages encrypted file storage on disk.
type FileStore struct {
	FilesDir string
	VaultKey []byte
}

// NewFileStore creates a new FileStore.
func NewFileStore(dataDir string, vaultKey []byte) *FileStore {
	dir := filepath.Join(dataDir, "files")
	os.MkdirAll(dir, 0700)
	return &FileStore{FilesDir: dir, VaultKey: vaultKey}
}

// StoreFile encrypts and stores a file and its manifest.
func (fs *FileStore) StoreFile(fileID string, fileData []byte, manifest *Manifest) error {
	fileKey, err := DeriveFileKey(fs.VaultKey, fileID)
	if err != nil {
		return err
	}
	encrypted, err := crypto.EncryptRaw(fileKey, fileData)
	if err != nil {
		return err
	}

	encPath := filepath.Join(fs.FilesDir, fileID+".enc")
	if err := os.WriteFile(encPath, encrypted, 0600); err != nil {
		return err
	}

	metaData, err := json.MarshalIndent(manifest, "", "  ")
	if err != nil {
		return err
	}
	metaPath := filepath.Join(fs.FilesDir, fileID+".meta")
	return os.WriteFile(metaPath, metaData, 0644)
}

// LoadFile decrypts and returns a stored file.
func (fs *FileStore) LoadFile(fileID string) ([]byte, error) {
	encPath := filepath.Join(fs.FilesDir, fileID+".enc")
	encrypted, err := os.ReadFile(encPath)
	if err != nil {
		return nil, fmt.Errorf("file %s not found in store", fileID)
	}
	fileKey, err := DeriveFileKey(fs.VaultKey, fileID)
	if err != nil {
		return nil, err
	}
	return crypto.DecryptRaw(fileKey, encrypted)
}

// LoadManifest loads a file's manifest entry.
func (fs *FileStore) LoadManifest(fileID string) (*Manifest, error) {
	metaPath := filepath.Join(fs.FilesDir, fileID+".meta")
	data, err := os.ReadFile(metaPath)
	if err != nil {
		return nil, fmt.Errorf("manifest %s not found", fileID)
	}
	var m Manifest
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, err
	}
	return &m, nil
}

// ListManifests lists all stored file manifests.
func (fs *FileStore) ListManifests() ([]*Manifest, error) {
	entries, err := os.ReadDir(fs.FilesDir)
	if err != nil {
		return nil, err
	}
	var manifests []*Manifest
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".meta") {
			data, err := os.ReadFile(filepath.Join(fs.FilesDir, e.Name()))
			if err != nil {
				continue
			}
			var m Manifest
			if err := json.Unmarshal(data, &m); err != nil {
				continue
			}
			manifests = append(manifests, &m)
		}
	}
	return manifests, nil
}

// HasFile checks if a file is stored locally.
func (fs *FileStore) HasFile(fileID string) bool {
	_, err := os.Stat(filepath.Join(fs.FilesDir, fileID+".enc"))
	return err == nil
}

// StoreOwnFile reads a file from disk, creates a signed manifest, encrypts and stores it.
func (fs *FileStore) StoreOwnFile(identitySK ed25519.PrivateKey, identityPK ed25519.PublicKey,
	filePath, description string) (*Manifest, error) {

	filename := filepath.Base(filePath)
	fileData, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}
	manifest, err := CreateManifest(identitySK, identityPK, filename, fileData, description)
	if err != nil {
		return nil, err
	}
	return manifest, fs.StoreFile(manifest.FileID, fileData, manifest)
}

File: python-client/p2pshare/storage/file_store.py
"""Encrypted file storage and manifest management.

Files are stored encrypted at rest:
  file_key = HKDF-SHA256(ikm=vault_key, salt=file_id_bytes, info="p2pshare-file-v1", len=32)
  .enc file = nonce (12B) || AES-256-GCM(file_key, plaintext) || tag (16B)
  .meta file = plaintext JSON manifest entry (owner signature verifiable without vault)
"""

import os
import json
import hashlib
import time

from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes

from p2pshare.crypto.keys import (
    b64url_encode,
    b64url_decode,
    public_key_to_bytes,
    peer_id_from_public_key,
)
from p2pshare.crypto.signing import canonical_json, sign, verify
from p2pshare.crypto.session import encrypt_raw, decrypt_raw
from p2pshare.errors import IntegrityError

FILE_KEY_INFO = b"p2pshare-file-v1"


def derive_file_key(vault_key: bytes, file_id: str) -> bytes:
    """Derive a per-file AES-256-GCM key from the vault key and file_id."""
    file_id_bytes = b64url_decode(file_id)
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=file_id_bytes,
        info=FILE_KEY_INFO,
    )
    return hkdf.derive(vault_key)


def compute_file_id(owner_id_bytes: bytes, filename: str, upload_timestamp: int) -> str:
    """Compute a file_id: base64url(SHA256(owner_id || utf8(filename) || timestamp_bytes))."""
    ts_bytes = upload_timestamp.to_bytes(8, "big")
    digest = hashlib.sha256(owner_id_bytes + filename.encode("utf-8") + ts_bytes).digest()
    return b64url_encode(digest)


def compute_file_hash(data: bytes) -> str:
    """Compute base64url(SHA256(data))."""
    return b64url_encode(hashlib.sha256(data).digest())


def create_manifest(identity_sk, identity_pk, filename: str, file_data: bytes, description: str = "") -> dict:
    """Create a signed file manifest entry for a file the local peer owns."""
    owner_id_bytes = public_key_to_bytes(identity_pk)
    upload_ts = int(time.time() * 1000)
    file_id = compute_file_id(owner_id_bytes, filename, upload_ts)
    file_hash = compute_file_hash(file_data)

    manifest = {
        "file_id": file_id,
        "filename": filename,
        "size_bytes": len(file_data),
        "sha256": file_hash,
        "owner_id": peer_id_from_public_key(identity_pk),
        "upload_timestamp": upload_ts,
        "description": description,
        "version": 1,
    }
    sig = sign(identity_sk, canonical_json(manifest))
    manifest["owner_signature"] = b64url_encode(sig)
    return manifest


def verify_manifest(manifest: dict, owner_pk) -> None:
    """Verify the owner_signature on a manifest entry.

    Raises IntegrityError if invalid.
    """
    sig_b64 = manifest["owner_signature"]
    sig = b64url_decode(sig_b64)

    manifest_without_sig = {k: v for k, v in manifest.items() if k != "owner_signature"}
    verify(owner_pk, canonical_json(manifest_without_sig), sig)


def verify_file_integrity(file_data: bytes, expected_hash_b64: str) -> None:
    """Verify a downloaded file's SHA-256 matches the manifest. Raises IntegrityError."""
    actual = compute_file_hash(file_data)
    if actual != expected_hash_b64:
        raise IntegrityError("Downloaded file is corrupted or was tampered with. File discarded.")


class FileStore:
    """Manages encrypted file storage on disk."""

    def __init__(self, data_dir: str, vault_key: bytes):
        self.files_dir = os.path.join(data_dir, "files")
        self.vault_key = vault_key
        os.makedirs(self.files_dir, exist_ok=True)

    def store_file(self, file_id: str, file_data: bytes, manifest: dict) -> None:
        """Encrypt and store a file and its manifest."""
        file_key = derive_file_key(self.vault_key, file_id)
        encrypted = encrypt_raw(file_key, file_data)

        enc_path = os.path.join(self.files_dir, f"{file_id}.enc")
        with open(enc_path, "wb") as f:
            f.write(encrypted)

        meta_path = os.path.join(self.files_dir, f"{file_id}.meta")
        with open(meta_path, "w") as f:
            json.dump(manifest, f, indent=2)

    def load_file(self, file_id: str) -> bytes:
        """Decrypt and return a stored file's plaintext content."""
        enc_path = os.path.join(self.files_dir, f"{file_id}.enc")
        if not os.path.exists(enc_path):
            raise FileNotFoundError(f"File {file_id} not found in store")

        file_key = derive_file_key(self.vault_key, file_id)
        with open(enc_path, "rb") as f:
            encrypted = f.read()
        return decrypt_raw(file_key, encrypted)

    def load_manifest(self, file_id: str) -> dict:
        """Load a file's manifest entry."""
        meta_path = os.path.join(self.files_dir, f"{file_id}.meta")
        if not os.path.exists(meta_path):
            raise FileNotFoundError(f"Manifest {file_id} not found")
        with open(meta_path, "r") as f:
            return json.load(f)

    def list_manifests(self) -> list[dict]:
        """List all stored file manifests."""
        manifests = []
        for name in os.listdir(self.files_dir):
            if name.endswith(".meta"):
                path = os.path.join(self.files_dir, name)
                with open(path, "r") as f:
                    manifests.append(json.load(f))
        return manifests

    def has_file(self, file_id: str) -> bool:
        """Check if a file is stored locally."""
        return os.path.exists(os.path.join(self.files_dir, f"{file_id}.enc"))

    def store_own_file(self, identity_sk, identity_pk, filepath: str, description: str = "") -> dict:
        """Read a file from disk, create a signed manifest, encrypt and store it.

        Returns the manifest entry.
        """
        filename = os.path.basename(filepath)
        with open(filepath, "rb") as f:
            file_data = f.read()

        manifest = create_manifest(identity_sk, identity_pk, filename, file_data, description)
        self.store_file(manifest["file_id"], file_data, manifest)
        return manifest
